### PR TITLE
Add L1 search

### PR DESCRIPTION
**Pull request by yannmclatchie**
## Description

This PR will add to ``kulprit`` the L1 search functionality developed to afford computationally cheaper projection search paths as is suggested by Piironen et al. (2020). We further remove the analytic solution features from the repo since the added weight and complexity of the code, and their future maintenance outweighs the benefit derived from them.

## Checklist

- [x] Tests covering the new functionality have been added
- [x] Documentation has been updated OR the change is too minor to be documented
- [ ] Changes are listed in the `CHANGELOG.md` OR changes are insignificant

### CODE REVIEW

File: src/kulprit/search/l1.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+        # initialise dictionary of indices
+        idx_dict = {}
+
+        # loop through each row and find first non-zero element
+        for i, j in zip(*np.where(arr > 0)):
+            if i in idx_dict:
+                continue
+            else:
+                idx_dict[i] = j
```
Comment (tomicapretto) [inline src/kulprit/search/l1.py:63]: Nitpick, but you don't need the `else` after the `continue`

File: src/kulprit/search/l1.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+        Returns:
+            dict: Dictionary keyed by the row number where each value is the index
+                of the first non-zero element in that row."""
+
+        # initialise dictionary of indices
+        idx_dict = {}
+
+        # loop through each row and find first non-zero element
+        for i, j in zip(*np.where(arr > 0)):
```
Comment (tomicapretto) [inline src/kulprit/search/l1.py:59]: I'm not sure what the matrix `arr` contains, but writing this to double check that you have 0 or positive values. If you can have negative values, you should use `np.where(arr != 0)`

**Comment by tomicapretto** [comment]
I agree that maintaining analytic solutions can be too much work and the benefits don't outweight the cost. However, and this is just an idea, would it be possible to allow users to extend kulprit so they can implement their own analytic solutions?

I imagine a scenario where kulprit does not provide analytic solutions, but allows you to register a method to do so. I may be proposing something too ambitious at this early stage of development, but since you already wrote all the code for the Gaussian family, it would be a pitty to just drop it.

Another drawback thta comes to my mind is that it may require too much knowledge of the internals, so people won't implement any analytic solution because of that complexity.

File: src/kulprit/search/l1.py (hunk 2 above)
Comment (yannmclatchie) [reply src/kulprit/search/l1.py:59]: You are absolutely right, thanks!

**Comment by yannmclatchie** [comment]
The idea of having user-defined ``Solver`` classes is nice, since users may wish to solve the KL minimisation problem many different ways. My only concerns are that currently what the ``AnalyticSolver`` classes do is vectorise a hard-coded solution to the KL optimisation for a family-link pairing, and since each family can accommodate multiple link functions, it might become laborious to code for either a user or a developer. 

Naturally, removing this features comes at increased computational cost. Initially, I had included this Gaussian code since it is the simplest and with the idea of having a ``method`` flag only. In [this other PR](https://github.com/yannmclatchie/kulprit/pull/24) I hope to make the optimisation sufficiently robust and flexible that the end user not feel hard done by, and indeed the truly interesting area of flexibility as I see it isn't necessarily how we solve the problem (the ``Solver`` classes in this case) but rather the nature of the problem (``Architecture`` and ``Loss`` classes). I would be open to the idea of allowing for user-defined definitions of such classes in the future, but perhaps not the core solving mechanism.

**Comment by tomicapretto** [comment]
Makes much sense!

**Pull request merged**